javascript/ql/src/semmle/javascript/CharacterEscapes.qll

```diff
@@ -0,0 +1,88 @@
+/**
+ * Provides classes for reasoning about character escapes in literals.
+ */
+
+import javascript
+
+module CharacterEscapes {
+  /**
+   * Provides sets of characters (as strings) with specific string/regexp characteristics.
+   */
+  private module Sets {
+    string ordinaryEscapeChars() { result = "fnrtvux0" }
+
+    string regexpCharRangeChars() { result = "bBcdDsSwW" }
+
+    string regexpBackreferenceChars() { result = "123456789k" }
+
+    string regexpMetaChars() { result = "^$*+?.()|{}[]-" }
+  }
+
+  /**
+   * Gets the `i'th` character of `raw`, which is preceeded by an uneven number of backslashes.
+   */
+  bindingset[raw]
+  string getAnEscapedCharacter(string raw, int i) {
+    result = raw.regexpFind("(?<=(^|[^\\\\])\\\\(\\\\{2}){0,10}).", _, i)
+  }
+
+  /**
+   * Gets a character in `n` that is preceeded by a single useless backslash.
+   *
+   * The character is the `i'th` character of `rawStringNode`'s raw string value.
+   */
+  string getAnIdentityEscapedCharacter(DataFlow::Node n, ASTNode rawStringNode, int i) {
+    exists(string raw, string escapableChars |
+      exists(string quote |
```

Comment:
Do I understand correctly that this `exists` is to ensure that e.g. `\"` is flagged inside `'Foo \"bar\" baz'`.
Is this issue even worth flagging?

Reply:
We wont flag such an issue with `js/useless-regexp-character-escape`, as neither quote has special regex semantics. 
But `getAnIdentityEscapedCharacter` is intended to be complete and independent of the any query purpose, and I think it is appropriate that the dual query in `javascript/ql/test/query-tests/Security/CWE-020/UselessCharacterEscape.ql` flags `'\"'` as a useless escape.